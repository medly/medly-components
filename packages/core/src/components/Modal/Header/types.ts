import { HTMLProps, WithThemeProp } from '@medly-components/utils';
import { ModalProps, ScrollState } from '../types';

export interface Props extends ModalProps {
    setHeaderHeight?: (height: number) => void;
    isSmallScreen?: boolean;
}
export interface Props extends HTMLProps<HTMLDivElement>, WithThemeProp {
    id: string;
    scrollState: ScrollState;
}
