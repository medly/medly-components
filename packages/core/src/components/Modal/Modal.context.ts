/* eslint-disable @typescript-eslint/no-empty-function */
import { createContext } from 'react';
import { ModalContextType } from './types';

// default values are for testing purposes
export const defaultValues: ModalContextType = {
    headerHeight: 120,
    setHeaderHeight: (): void => {},
    scrollState: {
        scrolledToTop: true,
        scrolledToBottom: false,
        scrollPosition: 120
    },
    dispatch: (): void => {},
    id: 'default-id',
    isSmallScreen: false
};

export const ModalContext = createContext<ModalContextType>(defaultValues);
