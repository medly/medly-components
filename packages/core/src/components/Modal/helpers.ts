export const getRemFromPx = (px: number) => (px ? px / 10 : 0);
