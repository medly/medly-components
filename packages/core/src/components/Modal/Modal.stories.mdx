import { Modal } from './Modal';
import { Preview, Story, Meta, Props } from '@storybook/addon-docs/blocks';
import * as stories from './Modal.stories';

<Meta title="Core" component={Modal} parameters={{ jest: ['Modal.test.tsx'] }} />

# Modal

The `Modal` components is a separate windows within an application that is overlaid on the primary window. The contents behind a modal dialog are inert, meaning you cannot interact with the content behind the dialog. It is a common user interface pattern that provides information or requires confirmation.
It consists of a Header, Content, and Actions. You can play with the component in the canvas tab.

```tsx
import { Modal } from '@medly-components/core';
import React, { useState } from 'react';

export const DummyComponent: React.FC = React.memo(() => {
    const [state, setState] = useState(false);

    const openModal = () => setState(true);
    const closeModal = () => setState(false);

    return (
        <>
            <Button onClick={openModal}>Open Modal</Button>
            <Modal open={state} onCloseModal={closeModal}>
                <Modal.Header>Add User</Modal.Header>
                <Modal.Content>
                    <Input id="name-input" type="text" fullWidth label="Name" placeholder="Enter your Name" />
                </Modal.Content>
                <Modal.Actions>
                    <Button variant="outlined">Add User</Button>
                </Modal.Actions>
            </Modal>
        </>
    );
});

DummyComponent.displayName = 'DummyComponent';
```

## Shadow Effect

The shadow effect apears on the Header & Actions section if you scroll through the content.

### Keyboard Support

The Modal component supports keyboard navigation when in focus.

| Key | Function        |
| --- | --------------- |
| Esc | Close the modal |

You can disable the closing of modal on Escape key press by passing `disableEscapeKeyDown` prop.

<Preview withToolbar>
    <Story name="Modal">{stories.Basic()}</Story>
</Preview>

## Props

This component offers the following props mentioned in the table below:

<Props of={Modal} />

## Modal Action Props

The Modal action prop is given in the table below:

<Props of={stories.ModalActionProps} />

## ModalTheme

This component offers the following themes mentioned in the table below:

<Props of={stories.ThemeInterface} />
