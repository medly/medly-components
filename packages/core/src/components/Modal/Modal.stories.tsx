import { defaultTheme, ModalTheme } from '@medly-components/theme';
import React, { useCallback, useState } from 'react';
import styled from 'styled-components';
import Button from '../Button';
import Input from '../Input';
import SingleSelect from '../SingleSelect';
import { ModalActionUserProps } from './Actions/types';
import { Modal } from './Modal';

const options = [
    { value: '0', label: 'Admin' },
    { value: '1', label: 'Pharmacist' },
    { value: '2', label: 'Covered Entity' }
];

export const ThemeInterface: React.FC<ModalTheme> = () => null;
ThemeInterface.defaultProps = {
    ...defaultTheme.modal
};

export const ModalActionProps: React.FC<ModalActionUserProps> = () => null;
ModalActionProps.defaultProps = {
    alignItems: 'right'
};

const Form = styled('form')`
    & > * {
        margin-bottom: 1.6rem;
    }
`;

export const ModalStoryWrapper = styled('div')<{ expand: boolean }>`
    height: ${({ expand }) => (expand ? '50rem' : '5rem')};
`;

export const Basic = () => {
    const [modalState, setModalState] = useState(false);

    const changeModalState = useCallback(() => setModalState(val => !val), []);

    return (
        <ModalStoryWrapper expand={modalState}>
            <Button onClick={changeModalState}>Click to Open</Button>
            <Modal open={modalState} onCloseModal={changeModalState}>
                <Modal.Header>Add User</Modal.Header>
                <Modal.Content>
                    <Form>
                        <SingleSelect label="Role" fullWidth options={options} value="0" />
                        <Input id="name-input" type="text" fullWidth label="Name" placeholder="Enter your Name" />
                        <Input
                            id="email-input"
                            type="email"
                            fullWidth
                            required
                            label="Email Address"
                            placeholder="Enter Email Address"
                            description="We will never share your email with anyone"
                        />
                        <Input
                            id="password-input"
                            type="password"
                            fullWidth
                            required
                            label="New Password"
                            placeholder="Enter New Password"
                        />
                        <Input
                            id="cpassword-input"
                            type="password"
                            fullWidth
                            required
                            label="Confirm Password"
                            placeholder="Enter Confirm Password"
                        />
                        <Input
                            id="tel-input"
                            type="tel"
                            pattern="[0-9]{3}-[0-9]{3}-[0-9]{4}"
                            fullWidth
                            label="Contact no."
                            placeholder="Enter your Contactno."
                        />
                    </Form>
                </Modal.Content>
                <Modal.Actions>
                    <Button variant="outlined">Add User</Button>
                </Modal.Actions>
            </Modal>
        </ModalStoryWrapper>
    );
};
