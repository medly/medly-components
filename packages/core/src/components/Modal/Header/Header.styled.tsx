import { breakpoints, media } from '@medly-components/utils';
import { rgba } from 'polished';
import styled, { css } from 'styled-components';
import { getRemFromPx } from '../helpers';
import { Props } from './types';

const getPosition = ({ scrollState }: Props) => {
    const { scrollPosition } = scrollState;
    const scrollPositionRem = getRemFromPx(scrollPosition);

    const isWithinThreshold = 6.6 - scrollPositionRem > 2.1;

    return css`
        position: ${isWithinThreshold ? 'relative' : 'fixed'};
        padding: ${({ theme }) =>
            `${isWithinThreshold ? theme.spacing.L2 : theme.spacing.S4} ${theme.spacing.L2} ${theme.spacing.S4} ${theme.spacing.S4}`};

        ${({ theme }) => media(breakpoints(theme.breakpoints).up('M'))`
            position: relative;
            padding: ${theme.spacing.M2} ${theme.spacing.L2} ${theme.spacing.S4} ${theme.spacing.M2};
        `}

        ${({ theme }) => `
            @media (min-width: ${theme.breakpoints.M.min}px) and (max-height: 700px) {
                transition: padding 100ms linear;
                padding: 
                    ${scrollState.scrolledToTop ? theme.spacing.M2 : theme.spacing.S4} ${theme.spacing.M2} ${theme.spacing.S4} ${
            theme.spacing.M2
        };
            }
        `}
    `;
};

const getBoxShadow = ({ scrollState, theme }: Props) => {
    const { scrollPosition } = scrollState;
    const scrollPositionRem = getRemFromPx(scrollPosition);

    return css`
        box-shadow: ${6.6 - scrollPositionRem < 2.1 && `0 1.8rem 1.6rem -1.6rem ${rgba(theme.colors.grey[400], 0.6)}`};

        ${({ theme }) => media(breakpoints(theme.breakpoints).up('M'))`
            box-shadow: ${!scrollState.scrolledToTop ? `0 1.8rem 1.6rem -1.6rem ${rgba(theme.colors.grey[400], 0.6)};` : ''}
        `}
    `;
};

export const Header = styled('div')<Props>`
    box-sizing: border-box;
    background-color: ${({ theme }) => theme.modal.backgroundColor};
    width: 100%;
    border-top-left-radius: 1.2rem;
    border-top-right-radius: 1.2rem;
    z-index: 10;
    color: ${({ theme }) => theme.modal.headerColor};
    ${getPosition}
    ${getBoxShadow}
`;
