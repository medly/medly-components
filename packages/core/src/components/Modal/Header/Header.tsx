import { isValidStringOrNumber, WithStyle } from '@medly-components/utils';
import React, { useContext, useEffect, useRef } from 'react';
import Text from '../../Text';
import { ModalContext } from '../Modal.context';
import * as Styled from './Header.styled';

const Component: React.FC = React.memo(({ children }) => {
    const { id, setHeaderHeight, isSmallScreen, scrollState } = useContext(ModalContext);
    const headerRef = useRef<HTMLDivElement>(null);

    useEffect(() => {
        headerRef.current && setHeaderHeight(headerRef.current.offsetHeight);
    }, [headerRef.current, isSmallScreen]);

    return (
        <Styled.Header id={`${id}-header`} ref={headerRef} scrollState={scrollState}>
            {React.Children.map(children, c => {
                return isValidStringOrNumber(c) ? <Text textVariant={isSmallScreen ? 'h3' : 'h2'}>{c}</Text> : c;
            })}
        </Styled.Header>
    );
});
Component.displayName = 'Header';
export const Header: React.FC & WithStyle = Object.assign(Component, { Style: Styled.Header });
