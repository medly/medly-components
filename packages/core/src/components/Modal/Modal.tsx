import { useCombinedRefs, useKeyPress, useWindowSize, WithStyle } from '@medly-components/utils';
import { FC, forwardRef, memo, MouseEvent, useCallback, useEffect, useLayoutEffect, useReducer, useRef, useState } from 'react';
import Actions from './Actions';
import CloseIcon from './CloseIcon';
import Content from './Content';
import Header from './Header';
import { ModalContext } from './Modal.context';
import { InnerContainerStyled, ModalBackgroundStyled } from './Modal.styled';
import ModalManager from './ModalManager';
import Popup from './Popup';
import { reducer } from './scrollStateReducer/scrollStateReducer';
import { ModalProps, ModalStaticProps } from './types';
import { useScrollState } from './useScrollState';

const manager = new ModalManager();

const Component: FC<ModalProps> = memo(
    forwardRef((props, ref) => {
        const {
                open,
                onCloseModal,
                overflowVisible,
                children,
                minWidth,
                shouldCloseOnOutsideClick,
                minHeight,
                disableEscapeKeyDown,
                ...restProps
            } = props,
            id = restProps.id || 'medly-modal',
            modalRef = useCombinedRefs<HTMLDivElement>(ref, useRef(null)),
            isEscPressed = useKeyPress('Escape', false, modalRef),
            innerContainerRef = useRef<HTMLDivElement>(null),
            [headerHeight, setHeaderHeight] = useState(0),
            [activeElement, setActiveElement] = useState<HTMLElement>(),
            [scrollState, dispatch] = useReducer(reducer, { scrolledToTop: true, scrolledToBottom: false, scrollPosition: 0 }),
            [shouldRender, setShouldRender] = useState(open),
            { width: windowWidth } = useWindowSize(),
            isSmallScreen = windowWidth < 768;

        const handleCloseModal = useCallback(() => {
                if (modalRef.current) manager.remove(modalRef.current);
                onCloseModal && onCloseModal();
            }, [onCloseModal, manager]),
            handleBackgroundClick = useCallback(
                (event: MouseEvent<HTMLDivElement>) =>
                    event.currentTarget === event.target && shouldCloseOnOutsideClick && handleCloseModal(),
                [shouldCloseOnOutsideClick, handleCloseModal]
            ),
            handleAnimationEnd = useCallback(() => {
                if (!open) {
                    setShouldRender(false);
                    document.body.style.overflow = 'unset';
                }
            }, [open]);

        useEffect(() => {
            if (open) {
                setShouldRender(true);
                document.body.style.overflow = 'hidden';
            }
            return () => {
                document.body.style.overflow = 'unset';
            };
        }, [open]);

        useEffect(() => {
            !disableEscapeKeyDown && open && isEscPressed && modalRef.current && manager.isTopModal(modalRef.current) && handleCloseModal();
        }, [open, isEscPressed, disableEscapeKeyDown, handleCloseModal, manager]);

        useLayoutEffect(() => {
            open && setActiveElement(document.activeElement as HTMLElement);
        }, [open]);

        useEffect(() => {
            !shouldRender && activeElement?.focus();
            if (shouldRender && modalRef.current) manager.add(modalRef.current);
        }, [shouldRender, manager]);

        const handleScroll = useScrollState({ ref: innerContainerRef, scrollState, dispatch });

        return shouldRender ? (
            <ModalBackgroundStyled {...{ ...restProps, id, open }} onClick={handleBackgroundClick}>
                <Popup
                    ref={modalRef}
                    id={`${id}-popup`}
                    onAnimationEnd={handleAnimationEnd}
                    {...{ minWidth, minHeight, open, overflowVisible }}
                >
                    <CloseIcon id={`${id}-close-button`} title={`${id}-close-icon`} size="M" variant="solid" onClick={handleCloseModal} />
                    <InnerContainerStyled
                        id={`${id}-inner-container`}
                        ref={innerContainerRef}
                        headerHeight={headerHeight}
                        overflowVisible={overflowVisible}
                        onScroll={handleScroll}
                    >
                        <ModalContext.Provider
                            value={{ headerHeight, setHeaderHeight, scrollState, dispatch, id, isSmallScreen, overflowVisible }}
                        >
                            {children}
                        </ModalContext.Provider>
                    </InnerContainerStyled>
                </Popup>
            </ModalBackgroundStyled>
        ) : null;
    })
);

Component.defaultProps = {
    open: false,
    shouldCloseOnOutsideClick: false,
    disableEscapeKeyDown: false
};
Component.displayName = 'Modal';
export const Modal: FC<ModalProps> & WithStyle & ModalStaticProps = Object.assign(Component, {
    Style: ModalBackgroundStyled,
    Header,
    Popup,
    Content,
    Actions
});
