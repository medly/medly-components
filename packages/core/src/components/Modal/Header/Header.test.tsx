import { render } from '@test-utils';
import { defaultValues, ModalContext } from '../Modal.context';
import { Header } from './Header';

const ModalHeaderRenderer = ({ isSmallScreen = true }: { isSmallScreen: boolean }) => {
    return render(
        <ModalContext.Provider value={{ ...defaultValues, isSmallScreen }}>
            <Header>test</Header>
        </ModalContext.Provider>
    );
};

describe('Modal component header', () => {
    it('should render properly at small screen size', () => {
        const { container } = ModalHeaderRenderer({ isSmallScreen: true });
        expect(container).toMatchSnapshot();
    });
});
