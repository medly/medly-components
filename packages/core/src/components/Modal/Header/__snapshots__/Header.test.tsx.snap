// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component header should render properly at small screen size 1`] = `
.c1 {
  margin: 0;
  color: inherit;
  font-size: 2.4rem;
  font-weight: 400;
  -webkit-letter-spacing: -0.04rem;
  -moz-letter-spacing: -0.04rem;
  -ms-letter-spacing: -0.04rem;
  letter-spacing: -0.04rem;
  line-height: 3.2rem;
  text-align: initial;
}

.c0 {
  box-sizing: border-box;
  background-color: #ffffff;
  width: 100%;
  border-top-left-radius: 1.2rem;
  border-top-right-radius: 1.2rem;
  z-index: 10;
  color: #13181D;
  position: fixed;
  padding: 1.6rem 5.6rem 1.6rem 1.6rem;
  box-shadow: 0 1.8rem 1.6rem -1.6rem rgba(176,188,200,0.6);
}

@media (min-width:768px) {
  .c0 {
    position: relative;
    padding: 3.2rem 5.6rem 1.6rem 3.2rem;
  }
}

@media (min-width:768px) and (max-height:700px) {
  .c0 {
    -webkit-transition: padding 100ms linear;
    transition: padding 100ms linear;
    padding: 3.2rem 3.2rem 1.6rem 3.2rem;
  }
}

@media (min-width:768px) {

}

<div>
  <div
    class="c0"
    id="default-id-header"
  >
    <h3
      class="c1"
    >
      test
    </h3>
  </div>
</div>
`;
